src/client/datascience/jupyter/kernels/cellExecutionQueue.ts

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { IDisposable } from 'monaco-editor';
+import { NotebookCell, NotebookCellRunState, NotebookEditor } from '../../../../../types/vscode-proposed';
+import { disposeAllDisposables } from '../../../common/helpers';
+import { traceInfo } from '../../../common/logger';
+import { createDeferred } from '../../../common/utils/async';
+import { noop } from '../../../common/utils/misc';
+import { traceCellMessage } from '../../notebook/helpers/helpers';
+import { INotebook } from '../../types';
+import { CellExecution, CellExecutionFactory } from './cellExecution';
+// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
+const vscodeNotebookEnums = require('vscode') as typeof import('vscode-proposed');
+
+/**
+ * A queue responsible for execution of cells.
+ * If this has not completed execution of the cells queued, we can continue to add more cells to this job.
+ * All cells queued using `runCells` are added to the queue and processed in order they were added/queued.
+ */
+export class CellExecutionQueue {
+    private readonly cellExecutions = new WeakMap<NotebookCell, CellExecution>();
+    private readonly queueOfCellsToExecute: CellExecution[] = [];
+    private readonly disposables: IDisposable[] = [];
+    private readonly completion = createDeferred<void>();
+    private cancelledOrCompletedWithErrors = false;
+    private startedRunningCells = false;
+    private chainedCellExecutionPromise: Promise<NotebookCellRunState | undefined> = Promise.resolve(undefined);
+    /**
+     * Whether all cells have completed processing or cancelled, or some completed & others cancelled.
+     * Even if this property is true, its possible there is still some cleanup remaining (or some promises waiting to be completed).
+     */
+    public get completed(): boolean {
```

Comment:
As this is class represents a queue, should this property be renamed to `isEmpty`, that feels more appropriate for the class, but would that convey the same meaning?

@rchiodo @IanMatthewHuff /cc

Reply:
IsEmpty does sound better but the comment above scares me. ```Even if this property is true, it's possible there is still some cleanup remaining```. Uh why is that is the first thing I'd ask. Can we fix it?

Reply:
Yeah, I noticed that it signals this to true at the start of the error or cancel, not when the cancel is completed.

Reply:
> Uh why is that is the first thing I'd ask. Can we fix it?

Unfortunately i don't see an easy way.
Lets assume we queue 10 cells, and something fails.
When there's a failure we have to do two things:
1. Notify users of this class that this class CANNOT be used anymore (has it has failed)
2. We have to cancel other cells (this is an async operation - updating cell status).

So, step 1 is crucial. If for instance a user runs another cell in the time being, we need to ensure it does not get added to this queue, as this queue is in an error state & we need to create a whole new queue.

I could instead add two properties `isEmpty` and `failed`.
If `failed=true`, we still have pending updates that require updates to cancel other cells. But the caller knows that this cannot be used.
The caller then uses the two properties instead of using the single property `completed`.

Hope this makes sense.

Reply:
Two properties sounds better to me :)